File: frontend/src/postprocessRailwayRows.ts
// postprocessRailwayRows.ts — жёсткая evidence-only постобработка

export type AssetKind = "track" | "switch";

export interface ParsedRow {
  location: string | null;
  assetKind: AssetKind | null;
  assetNumber: string | null;
  reference: string | null;
  defect: string | null;
  speedLimit: number | string | null;
  note: string | null;
  sourceText: string;
  rawDefect?: string | null;
  canonicalDefect?: string | null;
  normativeDecision?: string | null;
  warnings?: string[];
}

export interface RailwayDisplayRow {
  "Nп/п": number;
  "Местонахождение (перегон, станция)": string;
  "№ пути, стрелочного перевода": string;
  "Привязка (км,пк,м)": string;
  "Выявленная неисправность": string;
  "Ограничение скорости": string;
  "Примечание": string;
  "Исходный текст"?: string;
}

const DASH = "—";

const SYNTHETIC_NOTE_PATTERNS: RegExp[] = [
  /2288\s*р/i,
  /движение\s+закрыва(?:ется|ть|ют)/i,
  /закрытие\s+движения/i,
  /запрещается\s+движение/i,
  /ограничение\s+скорости/i,
];

const QUALIFIER_RULES: Array<{ pattern: RegExp; canonical: string }> = [
  {
    pattern: /\b(?:в|на)?\s*остри[ея]\s+остряка\b/gi,
    canonical: "в острие остряка",
  },
  {
    pattern: /\bпо\s+прямому\s+направлению\b/gi,
    canonical: "по прямому направлению",
  },
  {
    pattern: /\bпо\s+боковому\s+направлению\b/gi,
    canonical: "по боковому направлению",
  },
  {
    pattern: /\bна\s+крестовине\b/gi,
    canonical: "на крестовине",
  },
  {
    pattern: /\bна\s+усовике\b/gi,
    canonical: "на усовике",
  },
];

export function normalizeSpaces(input: string): string {
  return input
    .replace(/\s+/g, " ")
    .replace(/\s+([,.;:])/g, "$1")
    .trim();
}

export function normalizeAsrText(input: string): string {
  return normalizeSpaces(
    input
      .replace(/\bпусть\b/gi, "путь")
      .replace(/\bколи\b/gi, "колеи")
      .replace(/\bстр[.\s]*п[.\s]?\b/gi, "стрелочный перевод ")
      .replace(/\bострии\s+остряка\b/gi, "острие остряка")
  );
}

export function dashIfEmpty(value: string | null | undefined): string {
  if (!value) return DASH;
  const normalized = normalizeSpaces(value);
  return normalized.length > 0 ? normalized : DASH;
}

function toTitleCaseRu(value: string): string {
  return value
    .split(/\s+/)
    .map((word) =>
      word
        .split("-")
        .map((part) =>
          part ? part.charAt(0).toUpperCase() + part.slice(1).toLowerCase() : part
        )
        .join("-")
    )
    .join(" ");
}

export function countAssetMarkers(sourceText: string): number {
  const text = normalizeAsrText(sourceText);
  const matches = text.match(
    /\b(?:путь\s+\d+|стрелоч(?:ный|ного)\s+перевод(?:\s*(?:№|номер))?\s*\d+)\b/gi
  );
  return matches?.length ?? 0;
}

export function detectLocationFromSource(sourceText: string): string | null {
  const text = normalizeAsrText(sourceText);

  const match = text.match(
    /\bстанц(?:ия|ии)\s+(.+?)(?=\s+(?:путь\s+\d+|стрелоч(?:ный|ного)\s+перевод(?:\s*(?:№|номер))?\s*\d+)|$)/i
  );

  if (!match?.[1]) return null;

  const location = normalizeSpaces(match[1])
    .replace(/[.,;:]+$/g, "")
    .trim();

  return location ? toTitleCaseRu(location) : null;
}

export function detectAssetFromSource(
  sourceText: string
): { kind: AssetKind; number: string } | null {
  const text = normalizeAsrText(sourceText);

  const switchMatch = text.match(
    /\bстрелоч(?:ный|ного)\s+перевод(?:\s*(?:№|номер))?\s*(\d+)\b/i
  );
  if (switchMatch?.[1]) {
    return { kind: "switch", number: switchMatch[1] };
  }

  const trackMatch = text.match(/\bпуть\s+(\d+)\b/i);
  if (trackMatch?.[1]) {
    return { kind: "track", number: trackMatch[1] };
  }

  return null;
}

export function formatAssetForCell(
  assetKind: AssetKind | null,
  assetNumber: string | null
): string {
  if (!assetKind || !assetNumber) return DASH;
  return assetKind === "switch" ? `стр. п. ${assetNumber}` : assetNumber;
}

export function parseExplicitSpeed(sourceText: string): number | null {
  const text = normalizeAsrText(sourceText);

  let match = text.match(/\b(\d+)\s*км\/?\s*ч\b/i);
  if (match?.[1]) return Number(match[1]);

  match = text.match(
    /\b(?:ограничени[ея]\s+скорост[ьи]|скорост[ьи])\s*(?:до|не\s+более)?\s*(\d+)\b/i
  );
  if (match?.[1]) return Number(match[1]);

  return null;
}

export function containsExplicitSpeed(sourceText: string): boolean {
  return parseExplicitSpeed(sourceText) !== null;
}

function splitNoteParts(note: string): string[] {
  return note
    .replace(/[;]+/g, ".")
    .split(/[.]+/g)
    .map((part) => normalizeSpaces(part))
    .filter(Boolean);
}

function canonicalizeNotePart(part: string): string {
  let value = normalizeAsrText(part.toLowerCase())
    .replace(/[.,;:]+$/g, "")
    .trim();

  value = value
    .replace(/\bострие\s+остряка\b/g, "в острие остряка")
    .replace(/\bв\s+в\s+/g, "в ")
    .replace(/\bна\s+острие\s+остряка\b/g, "в острие остряка")
    .replace(/\bв\s+острии\s+остряка\b/g, "в острие остряка");

  return normalizeSpaces(value);
}

function dedupeNoteParts(parts: string[]): string | null {
  const result: string[] = [];
  const seen = new Set<string>();

  for (const part of parts) {
    const canonical = canonicalizeNotePart(part);
    if (!canonical) continue;
    if (seen.has(canonical)) continue;
    seen.add(canonical);
    result.push(canonical);
  }

  return result.length > 0 ? result.join(". ") : null;
}

function removeSyntheticNoteParts(note: string | null, sourceText: string): string | null {
  if (!note) return null;

  const source = normalizeAsrText(sourceText);
  const sourceContainsSynthetic = SYNTHETIC_NOTE_PATTERNS.some((re) => re.test(source));

  const filtered = splitNoteParts(note).filter((part) => {
    const isSynthetic = SYNTHETIC_NOTE_PATTERNS.some((re) => re.test(part));
    return sourceContainsSynthetic || !isSynthetic;
  });

  return dedupeNoteParts(filtered);
}

function extractQualifiersFromText(text: string): { cleanText: string; qualifiers: string[] } {
  let clean = normalizeSpaces(text);
  const qualifiers: string[] = [];

  for (const rule of QUALIFIER_RULES) {
    if (rule.pattern.test(clean)) {
      qualifiers.push(rule.canonical);
      clean = clean.replace(rule.pattern, " ");
      clean = normalizeSpaces(clean);
    }
    rule.pattern.lastIndex = 0;
  }

  clean = clean
    .replace(/\s+[.,;:]/g, "")
    .replace(/[.,;:]+$/g, "")
    .trim();

  return {
    cleanText: normalizeSpaces(clean),
    qualifiers,
  };
}

function stripLeadingContext(defect: string): string {
  let value = defect;

  value = value.replace(
    /\bстанц(?:ия|ии)\s+.+?(?=\s+(?:путь\s+\d+|стрелоч(?:ный|ного)\s+перевод(?:\s*(?:№|номер))?\s*\d+)|$)/i,
    " "
  );

  value = value.replace(
    /\b(?:путь\s+\d+|стрелоч(?:ный|ного)\s+перевод(?:\s*(?:№|номер))?\s*\d+)\b/gi,
    " "
  );

  return normalizeSpaces(value);
}

function normalizeDefectPunctuation(value: string): string {
  return normalizeSpaces(
    value
      .replace(/[;]+/g, " ")
      .replace(/\s+[.,:]/g, "")
      .replace(/[.,:]+$/g, "")
  );
}

function normalizeSleeperDefectLeak(defect: string): string {
  if (!/шпал/i.test(defect)) return defect;
  return normalizeSpaces(defect.replace(/\s+\d+\s*мм\b/gi, ""));
}

function normalizeLocationValue(
  explicitLocation: string | null | undefined,
  sourceText: string
): string | null {
  const fromRow = explicitLocation ? normalizeSpaces(explicitLocation) : "";
  if (fromRow) return toTitleCaseRu(fromRow);
  return detectLocationFromSource(sourceText);
}

function normalizeReferenceValue(reference: string | null | undefined): string | null {
  if (!reference) return null;
  const normalized = normalizeSpaces(reference);
  return normalized || null;
}

function extractDeterministicDefect(sourceText: string): {
  defect: string | null;
  note: string | null;
} {
  const text = normalizeAsrText(sourceText);

  let match = text.match(/износ\s+рамного\s+рельса\s+(\d+)\s*мм/i);
  if (match?.[1]) {
    const note = /\bостри[ея]\s+остряка\b/i.test(text) ? "в острие остряка" : null;
    return {
      defect: `износ рамного рельса ${match[1]} мм`,
      note,
    };
  }

  match = text.match(/ширин[аы]\s+коле[иы]\s+(\d+)\s*мм/i);
  if (match?.[1]) {
    return {
      defect: `ширина колеи ${match[1]} мм`,
      note: null,
    };
  }

  match = text.match(/(\d+)\s+подряд\s+куста\s+из\s+(\d+)\s+шпал/i);
  if (match?.[1] && match?.[2]) {
    return {
      defect: `${match[1]} подряд куста из ${match[2]} шпал`,
      note: null,
    };
  }

  match = text.match(/куст(?:\s+из)?\s+(\d+)\s+негодных\s+шпал/i);
  if (match?.[1]) {
    return {
      defect: `куст из ${match[1]} негодных шпал`,
      note: null,
    };
  }

  return { defect: null, note: null };
}

function formatSpeedForCell(speedLimit: number | string | null): string {
  if (speedLimit === null || speedLimit === undefined || speedLimit === "") return DASH;

  if (typeof speedLimit === "number") {
    return `${speedLimit} км/ч`;
  }

  const normalized = normalizeSpaces(String(speedLimit));
  if (!normalized) return DASH;

  if (/^\d+$/.test(normalized)) {
    return `${normalized} км/ч`;
  }

  return normalized;
}

export function sanitizeRowForExport(row: ParsedRow): ParsedRow {
  const source = normalizeAsrText(row.sourceText || "");
  const assetMarkerCount = countAssetMarkers(source);

  const warnings = [...(row.warnings ?? [])];

  if (assetMarkerCount > 1) {
    warnings.push(
      "sourceText содержит несколько объектов; для корректной постобработки нужен сегмент одной строки"
    );
  }

  const sourceLooksLikeSingleSegment = assetMarkerCount <= 1;

  const detectedAsset = sourceLooksLikeSingleSegment ? detectAssetFromSource(source) : null;
  const normalizedAssetKind = detectedAsset?.kind ?? row.assetKind ?? null;
  const normalizedAssetNumber = detectedAsset?.number ?? row.assetNumber ?? null;

  const normalizedLocation = normalizeLocationValue(row.location, source);
  const normalizedReference = normalizeReferenceValue(row.reference);

  const explicitSpeed = sourceLooksLikeSingleSegment ? parseExplicitSpeed(source) : null;
  const normalizedSpeed = explicitSpeed !== null ? explicitSpeed : null;

  const noteParts: string[] = [];

  const initialNote = removeSyntheticNoteParts(row.note ?? null, source);
  if (initialNote) {
    noteParts.push(...splitNoteParts(initialNote));
  }

  let defectFromRow = row.defect ? normalizeDefectPunctuation(stripLeadingContext(row.defect)) : null;

  if (defectFromRow) {
    const extracted = extractQualifiersFromText(defectFromRow);
    defectFromRow = extracted.cleanText || null;
    noteParts.push(...extracted.qualifiers);
  }

  let finalDefect = defectFromRow;

  if (sourceLooksLikeSingleSegment) {
    const deterministic = extractDeterministicDefect(source);

    if (deterministic.defect) {
      finalDefect = deterministic.defect;
    }

    if (deterministic.note) {
      noteParts.push(deterministic.note);
    }

    const gaugeMatch = source.match(/ширин[аы]\s+коле[иы]\s+(\d+)\s*мм/i);
    if (gaugeMatch?.[1]) {
      finalDefect = `ширина колеи ${gaugeMatch[1]} мм`;
    }
  }

  if (finalDefect) {
    finalDefect = normalizeDefectPunctuation(finalDefect);
    finalDefect = normalizeSleeperDefectLeak(finalDefect);
  }

  const finalNote = dedupeNoteParts(noteParts);

  return {
    ...row,
    location: normalizedLocation,
    assetKind: normalizedAssetKind,
    assetNumber: normalizedAssetNumber,
    reference: normalizedReference,
    defect: finalDefect || null,
    speedLimit: normalizedSpeed,
    note: finalNote,
    sourceText: source,
    rawDefect: row.rawDefect ?? row.defect ?? null,
    canonicalDefect: finalDefect || null,
    normativeDecision: null,
    warnings,
  };
}

export function sanitizeRowsForExport(rows: ParsedRow[]): ParsedRow[] {
  return rows.map(sanitizeRowForExport);
}

export function toDisplayRows(
  rows: ParsedRow[],
  options?: { includeSourceText?: boolean }
): RailwayDisplayRow[] {
  const sanitized = sanitizeRowsForExport(rows);

  return sanitized.map((row, index) => {
    const result: RailwayDisplayRow = {
      "Nп/п": index + 1,
      "Местонахождение (перегон, станция)": dashIfEmpty(row.location),
      "№ пути, стрелочного перевода": formatAssetForCell(row.assetKind, row.assetNumber),
      "Привязка (км,пк,м)": dashIfEmpty(row.reference),
      "Выявленная неисправность": dashIfEmpty(row.defect),
      "Ограничение скорости": formatSpeedForCell(row.speedLimit),
      "Примечание": dashIfEmpty(row.note),
    };

    if (options?.includeSourceText) {
      result["Исходный текст"] = dashIfEmpty(row.sourceText);
    }

    return result;
  });
}

File: backend/app/services/postprocess_railway_rows.py
"""
postprocessRailwayRows — жёсткая evidence-only постобработка ParsedRow → таблица.

Порт frontend/postprocessRailwayRows.ts
"""

from __future__ import annotations

import re
from typing import Any, Literal, TypedDict

from app.services.asr_fixes import normalize_asr_text

AssetKind = Literal["track", "switch"]
ParsedRow = dict[str, Any]

DASH = "—"

SYNTHETIC_NOTE_PATTERNS: tuple[re.Pattern[str], ...] = (
    re.compile(r"2288\s*р", re.IGNORECASE),
    re.compile(r"движение\s+закрыва(?:ется|ть|ют)", re.IGNORECASE),
    re.compile(r"закрытие\s+движения", re.IGNORECASE),
    re.compile(r"запрещается\s+движение", re.IGNORECASE),
    re.compile(r"ограничение\s+скорости", re.IGNORECASE),
)

QUALIFIER_RULES: tuple[tuple[re.Pattern[str], str], ...] = (
    (re.compile(r"\b(?:в|на)?\s*остри[ея]\s+остряка\b", re.IGNORECASE), "в острие остряка"),
    (re.compile(r"\bпо\s+прямому\s+направлению\b", re.IGNORECASE), "по прямому направлению"),
    (re.compile(r"\bпо\s+боковому\s+направлению\b", re.IGNORECASE), "по боковому направлению"),
    (re.compile(r"\bна\s+крестовине\b", re.IGNORECASE), "на крестовине"),
    (re.compile(r"\bна\s+усовике\b", re.IGNORECASE), "на усовике"),
)

_ASSET_MARKER_RE = re.compile(
    r"\b(?:путь\s+\d+|стрелоч(?:ный|ного)\s+перевод(?:\s*(?:№|номер))?\s*\d+)\b",
    re.IGNORECASE,
)
_STATION_LOCATION_RE = re.compile(
    r"\bстанц(?:ия|ии)\s+(.+?)(?=\s+(?:путь\s+\d+|стрелоч(?:ный|ного)\s+перевод(?:\s*(?:№|номер))?\s*\d+)|$)",
    re.IGNORECASE,
)
_SWITCH_ASSET_RE = re.compile(
    r"\bстрелоч(?:ный|ного)\s+перевод(?:\s*(?:№|номер))?\s*(\d+)\b",
    re.IGNORECASE,
)
_TRACK_ASSET_RE = re.compile(r"\bпуть\s+(\d+)\b", re.IGNORECASE)
_SPEED_KMH_RE = re.compile(r"\b(\d+)\s*км\s*/?\s*ч\b", re.IGNORECASE)
_SPEED_PHRASE_RE = re.compile(
    r"\b(?:ограничени[ея]\s+скорост[ьи]|скорост[ьи])\s*(?:до|не\s+более)?\s*(\d+)\b",
    re.IGNORECASE,
)
_STRIP_CONTEXT_STATION_RE = re.compile(
    r"\bстанц(?:ия|ии)\s+.+?(?=\s+(?:путь\s+\d+|стрелоч(?:ный|ного)\s+перевод(?:\s*(?:№|номер))?\s*\d+)|$)",
    re.IGNORECASE,
)
_STRIP_CONTEXT_ASSET_RE = re.compile(
    r"\b(?:путь\s+\d+|стрелоч(?:ный|ного)\s+перевод(?:\s*(?:№|номер))?\s*\d+)\b",
    re.IGNORECASE,
)
_GAUGE_DEFECT_RE = re.compile(r"ширин[аы]\s+коле[иы]\s+(\d+)\s*мм", re.IGNORECASE)
_MM_ON_SLEEPERS_RE = re.compile(r"\s+\d+\s*мм\b", re.IGNORECASE)

_WEAR_RE = re.compile(r"износ\s+рамного\s+рельса\s+(\d+)\s*мм", re.IGNORECASE)
_CLUSTER_RE = re.compile(r"(\d+)\s+подряд\s+куста\s+из\s+(\d+)\s+шпал", re.IGNORECASE)
_BAD_SLEEPERS_RE = re.compile(r"куст(?:\s+из)?\s+(\d+)\s+негодных\s+шпал", re.IGNORECASE)
_TIP_IN_SOURCE_RE = re.compile(r"\bостри[ея]\s+остряка\b", re.IGNORECASE)

MULTI_OBJECT_WARNING = (
    "sourceText содержит несколько объектов; "
    "для корректной постобработки нужен сегмент одной строки"
)


class RailwayDisplayRow(TypedDict, total=False):
    Nп_п: int  # not used — keys are Russian column names in runtime dicts


def normalize_spaces(input_text: str) -> str:
    text = re.sub(r"\s+", " ", input_text)
    text = re.sub(r"\s+([,.;:])", r"\1", text)
    return text.strip()


def dash_if_empty(value: str | None) -> str:
    if not value:
        return DASH
    normalized = normalize_spaces(value)
    return normalized if normalized else DASH


def _to_title_case_ru(value: str) -> str:
    words: list[str] = []
    for word in value.split():
        parts = []
        for part in word.split("-"):
            if part:
                parts.append(part[0].upper() + part[1:].lower())
            else:
                parts.append(part)
        words.append("-".join(parts))
    return " ".join(words)


def count_asset_markers(source_text: str) -> int:
    text = normalize_asr_text(source_text)
    return len(_ASSET_MARKER_RE.findall(text))


def detect_location_from_source(source_text: str) -> str | None:
    text = normalize_asr_text(source_text)
    match = _STATION_LOCATION_RE.search(text)
    if not match or not match.group(1):
        return None
    location = normalize_spaces(match.group(1))
    location = re.sub(r"[.,;:]+$", "", location).strip()
    return _to_title_case_ru(location) if location else None


def detect_asset_from_source(source_text: str) -> dict[str, str] | None:
    text = normalize_asr_text(source_text)
    switch_match = _SWITCH_ASSET_RE.search(text)
    if switch_match:
        return {"kind": "switch", "number": switch_match.group(1)}
    track_match = _TRACK_ASSET_RE.search(text)
    if track_match:
        return {"kind": "track", "number": track_match.group(1)}
    return None


def format_asset_for_cell(asset_kind: AssetKind | str | None, asset_number: str | None) -> str:
    if not asset_kind or not asset_number:
        return DASH
    if asset_kind == "switch":
        return f"стр. п. {asset_number}"
    return str(asset_number)


def parse_explicit_speed(source_text: str) -> int | None:
    text = normalize_asr_text(source_text)
    match = _SPEED_KMH_RE.search(text)
    if match:
        return int(match.group(1))
    match = _SPEED_PHRASE_RE.search(text)
    if match:
        return int(match.group(1))
    return None


def contains_explicit_speed(source_text: str) -> bool:
    return parse_explicit_speed(source_text) is not None


def _split_note_parts(note: str) -> list[str]:
    normalized = note.replace(";", ".")
    return [normalize_spaces(part) for part in normalized.split(".") if normalize_spaces(part)]


def _canonicalize_note_part(part: str) -> str:
    value = normalize_asr_text(part.lower())
    value = re.sub(r"[.,;:]+$", "", value).strip()
    value = re.sub(r"\bострие\s+остряка\b", "в острие остряка", value)
    value = re.sub(r"\bв\s+в\s+", "в ", value)
    value = re.sub(r"\bна\s+острие\s+остряка\b", "в острие остряка", value)
    value = re.sub(r"\bв\s+острии\s+остряка\b", "в острие остряка", value)
    return normalize_spaces(value)


def _dedupe_note_parts(parts: list[str]) -> str | None:
    result: list[str] = []
    seen: set[str] = set()
    for part in parts:
        canonical = _canonicalize_note_part(part)
        if not canonical or canonical in seen:
            continue
        seen.add(canonical)
        result.append(canonical)
    return ". ".join(result) if result else None


def _remove_synthetic_note_parts(note: str | None, source_text: str) -> str | None:
    if not note:
        return None
    source = normalize_asr_text(source_text)
    source_contains = any(p.search(source) for p in SYNTHETIC_NOTE_PATTERNS)
    filtered = [
        part
        for part in _split_note_parts(note)
        if source_contains or not any(p.search(part) for p in SYNTHETIC_NOTE_PATTERNS)
    ]
    return _dedupe_note_parts(filtered)


def _extract_qualifiers_from_text(text: str) -> tuple[str, list[str]]:
    clean = normalize_spaces(text)
    qualifiers: list[str] = []
    for pattern, canonical in QUALIFIER_RULES:
        if pattern.search(clean):
            qualifiers.append(canonical)
            clean = normalize_spaces(pattern.sub(" ", clean))
    clean = re.sub(r"\s+[.,;:]", "", clean)
    clean = re.sub(r"[.,;:]+$", "", clean).strip()
    return normalize_spaces(clean), qualifiers


def _strip_leading_context(defect: str) -> str:
    value = _STRIP_CONTEXT_STATION_RE.sub(" ", defect)
    value = _STRIP_CONTEXT_ASSET_RE.sub(" ", value)
    return normalize_spaces(value)


def _normalize_defect_punctuation(value: str) -> str:
    text = value.replace(";", " ")
    text = re.sub(r"\s+[.,:]", "", text)
    text = re.sub(r"[.,:]+$", "", text)
    return normalize_spaces(text)


def _normalize_sleeper_defect_leak(defect: str) -> str:
    if not re.search(r"шпал", defect, re.IGNORECASE):
        return defect
    return normalize_spaces(_MM_ON_SLEEPERS_RE.sub("", defect))


def _normalize_location_value(explicit_location: str | None, source_text: str) -> str | None:
    if explicit_location and normalize_spaces(explicit_location):
        return _to_title_case_ru(normalize_spaces(explicit_location))
    return detect_location_from_source(source_text)


def _normalize_reference_value(reference: str | None) -> str | None:
    if not reference:
        return None
    normalized = normalize_spaces(reference)
    return normalized or None


def extract_deterministic_defect(source_text: str) -> tuple[str | None, str | None]:
    text = normalize_asr_text(source_text)

    match = _WEAR_RE.search(text)
    if match:
        note = "в острие остряка" if _TIP_IN_SOURCE_RE.search(text) else None
        return f"износ рамного рельса {match.group(1)} мм", note

    match = _GAUGE_DEFECT_RE.search(text)
    if match:
        return f"ширина колеи {match.group(1)} мм", None

    match = _CLUSTER_RE.search(text)
    if match:
        return f"{match.group(1)} подряд куста из {match.group(2)} шпал", None

    match = _BAD_SLEEPERS_RE.search(text)
    if match:
        return f"куст из {match.group(1)} негодных шпал", None

    return None, None


def _format_speed_for_cell(speed_limit: int | str | None) -> str:
    if speed_limit is None or speed_limit == "":
        return DASH
    if isinstance(speed_limit, int):
        return f"{speed_limit} км/ч"
    normalized = normalize_spaces(str(speed_limit))
    if not normalized:
        return DASH
    if re.fullmatch(r"\d+", normalized):
        return f"{normalized} км/ч"
    return normalized


def sanitize_row_for_export(row: ParsedRow) -> ParsedRow:
    source = normalize_asr_text(str(row.get("sourceText") or ""))
    asset_marker_count = count_asset_markers(source)

    warnings: list[str] = list(row.get("warnings") or [])
    if asset_marker_count > 1 and MULTI_OBJECT_WARNING not in warnings:
        warnings.append(MULTI_OBJECT_WARNING)

    source_single = asset_marker_count <= 1

    detected = detect_asset_from_source(source) if source_single else None
    asset_kind = detected["kind"] if detected else row.get("assetKind")
    asset_number = detected["number"] if detected else row.get("assetNumber")

    location = _normalize_location_value(row.get("location"), source)
    reference = _normalize_reference_value(row.get("reference"))

    explicit_speed = parse_explicit_speed(source) if source_single else None
    speed_limit = explicit_speed if explicit_speed is not None else None

    note_parts: list[str] = []
    initial_note = _remove_synthetic_note_parts(
        str(row["note"]).strip() if row.get("note") else None,
        source,
    )
    if initial_note:
        note_parts.extend(_split_note_parts(initial_note))

    defect_from_row: str | None = None
    if row.get("defect"):
        defect_from_row = _normalize_defect_punctuation(_strip_leading_context(str(row["defect"])))
        extracted_clean, qualifiers = _extract_qualifiers_from_text(defect_from_row)
        defect_from_row = extracted_clean or None
        note_parts.extend(qualifiers)

    final_defect = defect_from_row

    if source_single:
        det_defect, det_note = extract_deterministic_defect(source)
        if det_defect:
            final_defect = det_defect
        if det_note:
            note_parts.append(det_note)
        gauge_match = _GAUGE_DEFECT_RE.search(source)
        if gauge_match:
            final_defect = f"ширина колеи {gauge_match.group(1)} мм"

    if final_defect:
        final_defect = _normalize_defect_punctuation(final_defect)
        final_defect = _normalize_sleeper_defect_leak(final_defect)

    final_note = _dedupe_note_parts(note_parts)

    return {
        **row,
        "location": location,
        "assetKind": asset_kind,
        "assetNumber": asset_number,
        "reference": reference,
        "defect": final_defect or None,
        "speedLimit": speed_limit,
        "note": final_note,
        "sourceText": source,
        "rawDefect": row.get("rawDefect") if row.get("rawDefect") is not None else row.get("defect"),
        "canonicalDefect": final_defect or None,
        "normativeDecision": None,
        "warnings": warnings,
    }


def sanitize_rows_for_export(rows: list[ParsedRow]) -> list[ParsedRow]:
    return [sanitize_row_for_export(row) for row in rows]


FORM_COLUMNS = (
    "Nп/п",
    "Местонахождение (перегон, станция)",
    "№ пути, стрелочного перевода",
    "Привязка (км,пк,м)",
    "Выявленная неисправность",
    "Ограничение скорости",
    "Примечание",
)


def to_display_rows(
    rows: list[ParsedRow],
    *,
    include_source_text: bool = False,
) -> tuple[list[str], list[dict[str, Any]]]:
    sanitized = sanitize_rows_for_export(rows)
    display: list[dict[str, Any]] = []
    for index, row in enumerate(sanitized, start=1):
        item: dict[str, Any] = {
            FORM_COLUMNS[0]: index,
            FORM_COLUMNS[1]: dash_if_empty(row.get("location")),
            FORM_COLUMNS[2]: format_asset_for_cell(row.get("assetKind"), row.get("assetNumber")),
            FORM_COLUMNS[3]: dash_if_empty(row.get("reference")),
            FORM_COLUMNS[4]: dash_if_empty(row.get("defect")),
            FORM_COLUMNS[5]: _format_speed_for_cell(row.get("speedLimit")),
            FORM_COLUMNS[6]: dash_if_empty(row.get("note")),
        }
        if include_source_text:
            item["Исходный текст"] = dash_if_empty(row.get("sourceText"))
        display.append(item)
    return list(FORM_COLUMNS), display


def parsed_row_from_record(rec: Any) -> ParsedRow:
    """ParsedRecord / FlatInspectionRow → ParsedRow."""
    asset_kind: str | None = None
    asset_number: str | None = None
    put = getattr(rec, "put", None)
    switch = getattr(rec, "switch", None)
    if put:
        asset_kind = "track"
        asset_number = str(put).strip()
    elif switch:
        asset_kind = "switch"
        asset_number = str(switch).strip()

    reference_parts: list[str] = []
    km = getattr(rec, "km", None)
    piket = getattr(rec, "piket", None)
    if km:
        reference_parts.append(f"{km} км")
    if piket:
        reference_parts.append(f"пк {piket}")

    speed = getattr(rec, "speed_limit", None)
    speed_limit: int | str | None = None
    if speed is not None and str(speed).strip():
        try:
            speed_limit = int(str(speed).strip())
        except ValueError:
            speed_limit = str(speed).strip()

    return {
        "location": getattr(rec, "uchastok", None) or getattr(rec, "peregon", None),
        "assetKind": asset_kind,
        "assetNumber": asset_number,
        "reference": ", ".join(reference_parts) or None,
        "defect": getattr(rec, "defect", None),
        "speedLimit": speed_limit,
        "note": getattr(rec, "comment", None),
        "sourceText": getattr(rec, "raw_text", None) or "",
        "warnings": [],
    }


# TS API aliases
sanitizeRowForExport = sanitize_row_for_export
sanitizeRowsForExport = sanitize_rows_for_export
toDisplayRows = to_display_rows
